### PR TITLE
Enhancement/2092 advanced analytics tracking fixes

### PR DESCRIPTION
**Pull request by eugene-manuilov**
## Summary

Addresses issue #2092

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: assets/js/analytics-advanced-tracking.js [hunk 1]
```diff
@@ -33,11 +32,10 @@ import setUpAdvancedTracking from './analytics-advanced-tracking/set-up-advanced
  * @param {Object} metadata Additional event metadata to send, or `null`.
  */
 function sendEvent( action, metadata ) {
-	if ( ! metadata ) {
-		global.gtag( 'event', action );
-	} else {
-		global.gtag( 'event', action, metadata );
-	}
+	window.gtag( 'event', action, metadata || undefined ); // eslint-disable-line no-restricted-globals
```
Comment (felixarntz) [inline assets/js/analytics-advanced-tracking.js:35]: Why not `global` like elsewhere?

Comment (eugene-manuilov) [reply assets/js/analytics-advanced-tracking.js:35]: @felixarntz it prevents webpack module concatenation optimization if we use global variables.

File: assets/js/analytics-advanced-tracking/set-up-advanced-tracking.js [hunk 2]
```diff
@@ -48,7 +48,7 @@ export default function setUpAdvancedTracking( eventConfigurations, sendEvent )
 
 	return () => {
 		toRemove.forEach( ( listenerArgs ) => {
-			document.removeEventListener( ...listenerArgs );
+			document.removeEventListener.apply( document, listenerArgs );
```
[the remark this answers is not in the document]
Comment (eugene-manuilov) [reply assets/js/analytics-advanced-tracking/set-up-advanced-tracking.js:51]: There are no technical differences, but if we use the old way to invoke the function with an array of arguments, babel won't try to inject a bunch of functions to transpile spread operation into an approach supported by old browsers.

**Comment by eugene-manuilov** [comment]
@felixarntz with all these changes I have managed to reduce the bundle size for advanced Analytics events tracking from 2.8kb to 1.6kb (~43%).

**Review by aaemnnosttv — DISMISSED** [review]
This works well for me although I'm seeing another problem which is that there is a difference in the number of triggers generated for web tracking than for AMP. This is because duplicate events may be added by a hook which runs more than once for the same hook.

I tested this locally using an `Event_List` that was included with tests that hooks 2 events onto `the_post` action. With the TwentyTwenty theme and a single post on the home page, it was run 5 times, which added a total of 10 triggers in "web mode".

For AMP however, event triggers are keyed with a hash based on the action, event, and selector. This resulted in only 2 triggers being registered when loading the same page in an AMP context.

It seems like we should see consistent results in the number of triggers/eventListeners configured regardless of AMP mode.

Ideally I think the unique handling would be done when registering the Event. Currently, this doesn't happen until the events are injected which means many duplicate `Event` instances may be accumulated unnecessarily.

@felixarntz what do you think about keying the added/registered `Events` in the array in the same way for both? That way if a duplicate `Event` is registered, it simply overwrites the previous in the array, and we can always get them all as a list with `array_values` similar to how we handle errors in the datastore.

**Comment by felixarntz** [comment]
@aaemnnosttv Good point, keying the events in the same way like in AMP makes sense.

**Comment by aaemnnosttv** [comment]
@felixarntz what do you think about adding a `Event::get_hash` method?

Then we can change `add_event` like so:

```php
	protected function add_event( Event $event ) {
		$this->events[ $event->get_hash() ] = $event;
	}

// alternatively, perhaps we just hash it in `add_event`?

	protected function add_event( Event $event ) {
		$hash = md5( wp_json_encode( $event->jsonSerialize() ) ); 
		$this->events[ $hash ] = $event;
	}
```

Then the injectors only need to iterate over the map of events which will already have the hash for each, and the injector can be more specific to passing the list along to the frontend.

**Comment by aaemnnosttv** [comment]
@googlebot I consent.

**Comment by aaemnnosttv** [comment]
@felixarntz I can confirm that these changes result in only the 2 expected event triggers in AMP and non-amp now:

image

image

File: includes/Modules/Analytics/Advanced_Tracking/AMP_Config_Injector.php [hunk 3]
```diff
@@ -37,11 +37,9 @@ public function inject_event_configurations( $gtag_amp_opt, $events ) {
 			$gtag_amp_opt['triggers'] = array();
 		}
 
-		foreach ( $events as $event ) {
+		foreach ( $events as $amp_trigger_key => $event ) {
```
Comment (felixarntz) [inline includes/Modules/Analytics/Advanced_Tracking/AMP_Config_Injector.php:40]: Strictly saying, we should maybe call this `$identifier` or `$key`, it' not specific for AMP. But not a big deal.

**Review by felixarntz — APPROVED** [review]
Awesome! Just one super-minor comment, but nothing blocking, so it's good to merge.

**Pull request merged**